src/SourceTerms/LightBulb/CPU_Src_Lightbulb.cpp

```diff
@@ -0,0 +1,403 @@
+
+} // FUNCTION : Src_SetAuxArray_Lightbulb
+#endif // #ifndef __CUDACC__
+
+
+
+// ======================================
+// II. Implement the source-term function
+// ======================================
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Src_Lightbulb
+// Description :  Apply the lightbulb scheme for neutrino heating and cooling
+//
+// Note        :  1. Invoked by CPU/GPU_SrcSolver_IterateAllCells()
+//                2. See Src_SetAuxArray_Lightbulb() for the values stored in AuxArray_Flt/Int[]
+//                3. Shared by both CPU and GPU
+//                4. Ref: Sean M. Couch, 2013, ApJ, 765, 29 (arXiv: 1206.4724), sec. 2
+//
+// Parameter   :  fluid             : Fluid array storing both the input and updated values
+//                                    --> Including both active and passive variables
+//                B                 : Cell-centered magnetic field
+//                SrcTerms          : Structure storing all source-term variables
+//                dt                : Time interval to advance solution
+//                dh                : Grid size
+//                x/y/z             : Target physical coordinates
+//                TimeNew           : Target physical time to reach
+//                TimeOld           : Physical time before update
+//                                    --> This function updates physical time from TimeOld to TimeNew
+//                MinDens/Pres/Eint : Density, pressure, and internal energy floors
+//                EoS               : EoS object
+//                AuxArray_*        : Auxiliary arrays (see the Note above)
+//
+// Return      :  fluid[]
+//-----------------------------------------------------------------------------------------
+GPU_DEVICE_NOINLINE
+static void Src_Lightbulb( real fluid[], const real B[],
+                           const SrcTerms_t *SrcTerms, const real dt, const real dh,
+                           const double x, const double y, const double z,
+                           const double TimeNew, const double TimeOld,
+                           const real MinDens, const real MinPres, const real MinEint,
+                           const EoS_t *EoS, const double AuxArray_Flt[], const int AuxArray_Int[] )
+{
+
+// check
+#  ifdef GAMER_DEBUG
+   if ( AuxArray_Flt == NULL )   printf( "ERROR : AuxArray_Flt == NULL in %s !!\n", __FUNCTION__ );
+   if ( AuxArray_Int == NULL )   printf( "ERROR : AuxArray_Int == NULL in %s !!\n", __FUNCTION__ );
+#  endif
+
+
+   const real Kelvin2MeV = AuxArray_Flt[SRC_AUX_KELVIN2MEV];
+   const real sEint2Code = AuxArray_Flt[SRC_AUX_VSQR2CODE ];
+
+   const real x0         = x - SrcTerms->BoxCenter[0];
+   const real y0         = y - SrcTerms->BoxCenter[1];
+   const real z0         = z - SrcTerms->BoxCenter[2];
+   const real r2_CGS     = SQR( SrcTerms->Unit_L ) * (  SQR( x0 ) + SQR( y0 ) + SQR( z0 )  );
```

Comment:
If I understood correctly, x0, y0, and z0 are box centers, r2_CGS should be the distance^2 to the box center. Should n't `r2_CGS= SQR(x-x0) + SQR(y-y0) + SQR(z-z0)`?